README.md
# timeline_editor

Early version of a timeline editor. Support:
* Move of element
* Context menu
* Zoom of timeline
* Progress indicator
* track boxes
* continuous tracks
* scrub

<img src="https://raw.githubusercontent.com/neckaros/timeline_editor/master/example/demo.gif" height="700" />

## Usage

### Installation

Add `timeline_editor` as a dependency in your pubspec.yaml file ([what?](https://pub.dev/packages/timeline_editor#-installing-tab-)).


### Import

Import timeline_editor:
```dart
import 'package:timeline_editor/timeline_editor.dart';
```

### Migrate from V2 to V3
We now use Duration instead of seconds. To migrate you can use the helper functions to transform your seconds to Duration:
```dart
Duration durationFromSeconds(double seconds)
double durationToSeconds(Duration duration)
```

### widgets
See [example](https://github.com/neckaros/timeline_editor/blob/master/example/lib/main.dart)
```dart

TimelineEditor(
                  positionStream: positionStream,
                  onPositionTap: (s) => position = s,
                  countTracks: 2,
                  trackBuilder: (track, pps, duration) => track == 1
                      ? TimelineEditorTrack(
                          defaultColor: Colors.green[700],
                          boxes: [
                            TimelineEditorBox(box1Start, 100,
                                onMoved: updateBox1,
                                color: Colors.blue,
                                onMovedEnd: () => print('end moved')),
                            TimelineEditorBox(157, 80),
                          ],
                          pixelsPerSeconds: pps,
                          duration: duration,
                        )
                      : TimelineEditorTrack.fromContinuous(
                          continuousBoxes:[
                            TimelineEditorContinuousBox(
                                Duration.zero,
                                color: Colors.deepOrange,
                                child: const Image(image: const AssetImage('assets/image2.jpg')),
                            ),
                            TimelineEditorContinuousBox(
                                box2Start,
                                menuEntries: [
                                PopupMenuItem<String>(child: Text('Delete'), value: 'deleted')
                                ],
                                onMoved: updateBox2,
                                onSelectedMenuItem: (v) {
                                print('Selected: $v');
                                setState(() {
                                    deleted = true;
                                });
                                },
                                onTap: (start, duration) =>
                                    print('tapped for $start to ${start + duration}'),
                                color: Colors.black,
                                child: const Image(image: const AssetImage('assets/image.jpg')),
                            ),
                            ],
                          pixelsPerSeconds: pps,
                          duration: duration,
                        ),
                  duration: Duration(seconds: 300),
                ))
```
## TimelineEditor
Main widget used to hold your tracks, display position and time

## TimelineEditorTrack
A provided track to use with the **TimelineEditor** track builder (you can use your own)

It can either hold simple **TimelineEditorBox** with a start and an end

or TimelineEditorTrack.fromContinuous with **TimelineEditorContinuousBox** to have a conitunuous track where you only set start time and duration will be calculated so the box go up to the next box


### Example

[/example](https://github.com/neckaros/timeline_editor/tree/master/example)


CHANGELOG.md
## 0.4.0-preview.1
First preview version of the rewrite for better performance and:
* Header for tracks
* custom icons and
Huge thanks to @argorain and @ritheshSalyan

## 0.3.4
### Fix
* Fix too much logs ([Issue 5](https://github.com/neckaros/timeline_editor/issues/5))
### Enhancements
* You can now provide your custom time widget ([Request 3](https://github.com/neckaros/timeline_editor/issues/3)):
```dart
TimelineEditor(
  timeWidgetExtent: 100,
  timeWidgetBuilder: (d, t) => Padding(
            padding: const EdgeInsets.only(left: 8.0),
            child: Text(
              '${d.inSeconds}/${t.inSeconds}',
              overflow: TextOverflow.ellipsis,
            ),
          ),
```

## 0.3.2
* Fix scroll will overflow sometimes

## 0.3.1
* Added helper to go from seconds double to Duration:
```dart
Duration durationFromSeconds(double seconds)
double durationToSeconds(Duration duration)
```
* Added possibility to customize timeline text theme and separator color:
```dart
class TimelineEditor extends StatefulWidget {
  ...
  /// Timeline time text theme. By default we use Theme.of(context).textTheme.bodyText1
  final TextTheme timelineTextStyle;

  /// Color used by the time separator in the timeline.
  /// By default we use Theme.of(context).brightness == Brightness.dark ? Colors.white60 : Colors.black87
  final Color separatorColor;
  ...
}
```

## 0.3.0
* We now use duration instead of seconds ([Issue 1](https://github.com/neckaros/timeline_editor/issues/1))

## 0.2.2
* better time indicators
* height of track is not customizable

## 0.2.1
* fix scaling


## 0.2.0
Performance optimization
* position now takes a stream to rebuild as little as possible
* layout builder replaced by a cached version to avoid rebuilding too often
* you can now react to user clicking a time on the timeline ruler

## 0.1.0

* First release


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:timeline_editor/timeline_editor.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  Duration totalDuration = Duration(seconds: 600);

  Duration box1Start = Duration(seconds: 30);
  Duration box1Duration = Duration(seconds: 50);
  bool box1Selected = false;
  Duration box2Start = Duration(seconds: 100);
  Duration box2Duration = Duration(seconds: 180);
  bool box2Selected = false;

  Duration box1bStart = Duration(seconds: 30);
  Duration box1bDuration = Duration(seconds: 50);
  bool box1bSelected = false;
  Duration box2bStart = Duration(seconds: 100);
  Duration box2bDuration = Duration(seconds: 180);
  bool box2bSelected = false;

  bool deleted = false;
  double position = 0;
  bool customTimeString = false;
  bool withHeaders = false;
  StreamController<double> positionController;
  Timer progressTimer;
  TimelineEditorScaleController scaleController;

  double scale;

  double _trackHeight = 100;

  AnimationController _controller;
  Animation<double> _animation;

  void moveBox1(Duration newStart) {
    if (box1Start + newStart < Duration.zero) {
      newStart = Duration.zero;
    }
    if (box1Start + newStart + box1Duration < box2Start)
      setState(() {
        box1Start += newStart;
      });
  }

  void moveBox1End(Duration move) {
    if (box1Start + box1Duration + move < box2Start)
      setState(() => box1Duration = box1Duration + move);
  }

  void moveBox2(Duration startMove) {
    var newStart = box2Start + startMove;
    if (box1Start + box1Duration < newStart &&
        newStart + box2Duration < totalDuration)
      setState(() {
        box2Start = newStart;
      });
  }

  void moveBox2End(Duration move) {
    if (box2Start + box2Duration + move < totalDuration)
      setState(() => box2Duration = box2Duration + move);
  }

  // 2nd track

  void moveBox1b(Duration newStart) {
    if (box1bStart + newStart < Duration.zero) {
      newStart = Duration.zero;
    }
    if (box1bStart + newStart + box1bDuration < box2bStart)
      setState(() {
        box1bStart += newStart;
      });
  }

  void moveBox1bEnd(Duration move) {
    if (box1bStart + box1bDuration + move < box2bStart)
      setState(() => box1bDuration = box1bDuration + move);
  }

  void moveBox2b(Duration startMove) {
    var newStart = box2bStart + startMove;
    if (box1Start + box1bDuration < newStart &&
        newStart + box2bDuration < totalDuration)
      setState(() {
        box2bStart = newStart;
      });
  }

  void moveBox2bEnd(Duration move) {
    if (box2bStart + box2bDuration + move < totalDuration)
      setState(() => box2bDuration = box2bDuration + move);
  }

  void positionUpdate(Timer timer) {
    position += 0.350;
    if (position > 300) position = 0;
    positionController.add(position);
    /* if (position + 0.1 < 300)
      Timer(Duration(milliseconds: 100), () => positionUpdate());*/
  }

  @override
  void initState() {
    super.initState();
    scaleController = TimelineEditorScaleController();
    positionController = StreamController<double>.broadcast();
    progressTimer = Timer.periodic(Duration(milliseconds: 350), positionUpdate);
    _controller =
        AnimationController(vsync: this, duration: Duration(seconds: 2));
    _animation = Tween(begin: 100.0, end: 200.0).animate(_controller)
      ..addListener(() {
        setState(() => _trackHeight = _animation.value);
      });
    positionUpdate(null);
  }

  @override
  void dispose() {
    scaleController.dispose();
    progressTimer?.cancel();
    positionController?.close();

    _controller?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      darkTheme: ThemeData.dark(),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Timeline_editor example app'),
        ),
        body: Column(
          children: <Widget>[
            Expanded(
                child: Center(
                    child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                const Text('Timeline_editor example app'),
                StreamBuilder<double>(
                    stream: scaleController.scaleUpdates,
                    initialData: 1,
                    builder: (context, snapshot) {
                      return Text('Current scale: ' + snapshot.data.toString());
                    }),
                RaisedButton(
                  child: const Text('Change track height'),
                  onPressed: () =>
                      _controller.status == AnimationStatus.forward ||
                              _controller.status == AnimationStatus.completed
                          ? _controller.reverse()
                          : _controller.forward(),
                ),
                SwitchListTile(
                    title: Text('Custom time string'),
                    value: customTimeString,
                    onChanged: (value) =>
                        setState(() => customTimeString = value)),
                SwitchListTile(
                    title: Text('Headers'),
                    value: withHeaders,
                    onChanged: (value) => setState(() => withHeaders = value)),
              ],
            ))),
            Padding(
                padding: const EdgeInsets.only(bottom: 8.0),
                child: TimelineEditor(
                  scaleController: scaleController,
                  minimumTimeWidgetExtent: customTimeString ? 100 : null,
                  leadingWidgetBuilder: withHeaders
                      ? (index) => Center(
                          child: RaisedButton(
                              onPressed: () {}, child: Text("$index")))
                      : null,
                  timelineLeadingWidget:
                      withHeaders ? Center(child: Text("HEADER")) : null,
                  timeWidgetBuilder: customTimeString
                      ? (d, t) => Padding(
                            padding: const EdgeInsets.only(left: 8.0),
                            child: Text(
                              '${d.inSeconds}/${t.inSeconds}',
                              overflow: TextOverflow.ellipsis,
                            ),
                          )
                      : null,
                  onPositionTap: (s) => position = s,
                  positionStream: positionController.stream,
                  countTracks: 2,
                  trackBuilder: (track, pps, duration, scrollControllers) =>
                      track == 0
                          ? TimelineEditorTrack(
                              // key: Key('separated'),
                              scrollControllers: scrollControllers,
                              defaultColor: Colors.green[700],
                              boxes: [
                                TimelineEditorCard(
                                  box1Start,
                                  duration: box1Duration,
                                  selected: box1Selected,
                                  onTap: () => setState(
                                      () => box1Selected = !box1Selected),
                                  color: Colors.blue,
                                  onMovedDuration: moveBox1End,
                                  onMovedStart: moveBox1,
                                ),
                                TimelineEditorCard(box2Start,
                                    duration: box2Duration,
                                    menuEntries: [
                                      PopupMenuItem<String>(
                                          child: Text('Demo!'), value: 'demo')
                                    ],
                                    onSelectedMenuItem: (v) {
                                      print(v);
                                    },
                                    onMovedDuration: moveBox2End,
                                    onMovedStart: moveBox2,
                                    selected: box2Selected,
                                    onTap: () => setState(
                                        () => box2Selected = !box2Selected),
                                    color: Colors.green),
                              ],
                              pixelsPerSeconds: pps,
                              duration: duration,
                            )
                          : TimelineEditorTrack(
                              // key: Key('separated'),
                              scrollControllers: scrollControllers,
                              defaultColor: Colors.green[700],
                              boxes: [
                                TimelineEditorCard(
                                  box1Start,
                                  duration: box1bDuration,
                                  selected: box1bSelected,
                                  onTap: () => setState(
                                      () => box1bSelected = !box1bSelected),
                                  color: Colors.red,
                                  onMovedDuration: moveBox1bEnd,
                                  onMovedStart: moveBox1b,
                                ),
                                TimelineEditorCard(box2bStart,
                                    duration: box2bDuration,
                                    borderColor: Colors.yellow,
                                    menuEntries: [
                                      PopupMenuItem<String>(
                                          child: Text('Demo 2!'), value: 'demo')
                                    ],
                                    onSelectedMenuItem: (v) {
                                      print(v);
                                    },
                                    onMovedDuration: moveBox2bEnd,
                                    onMovedStart: moveBox2b,
                                    selected: box2bSelected,
                                    onTap: () => setState(
                                        () => box2bSelected = !box2bSelected),
                                    color: Colors.deepOrange),
                              ],
                              pixelsPerSeconds: pps,
                              duration: duration,
                            ),
                  duration: totalDuration,
                )),
          ],
        ),
      ),
    );
  }
}
